src/gc.rs

```diff
@@ -39,41 +39,99 @@ use crate::GC_ALLOCATOR;
 /// free `Gc` values behind the scenes.
 #[derive(PartialEq, Eq, Debug)]
 pub struct Gc<T: ?Sized> {
-    pub(crate) objptr: NonNull<GcBox<T>>,
+    pub(crate) ptr: NonNull<GcBox<T>>,
+    _phantom: PhantomData<T>,
 }
 
 impl<T> Gc<T> {
     /// Constructs a new `Gc<T>`.
     pub fn new(v: T) -> Self {
         Gc {
-            objptr: unsafe { NonNull::new_unchecked(GcBox::new(v)) },
+            ptr: unsafe { NonNull::new_unchecked(GcBox::new(v)) },
+            _phantom: PhantomData,
         }
     }
+
+    /// Constructs a new `Gc<MaybeUninit<T>>` which is capable of storing data
+    /// up-to the size permissible by `layout`.
+    ///
+    /// This can be useful if you want to store a value with a custom layout,
+    /// but have the collector treat the value as if it were T.
+    ///
+    /// If `layout` is insufficiently sized to store `T`, a Gc value will not be
+    /// allocated. This is because it's possible to accidentally access out-of-
+    /// bounds memory when treating the Gc value as if it contains a type larger
+    /// than its actual size.
+    pub fn new_from_layout(layout: Layout) -> Option<Gc<MaybeUninit<T>>> {
+        if layout.size() < Layout::new::<T>().size() {
```

Comment:
Need to check for alignment too.

Reply:
Fixed: 93953995086dc4076967b69f5726ba4d29b14ffd